### PR TITLE
Support OpenSSL 1.1.1

### PR DESCRIPTION
**Pull request by ysbaddaden**
OpenSSL 1.1.0 introduced a public API to create and populate the BIO and BIO_METHOD structs. We now use it when available. We still keep the manual creation for OpenSSL 1.0.2 (and below) which is still supported.

This fixes an internal ABI breaking change introduced in OpenSSL 1.1.1 that causes a compiled program to segfault. See #6737.

This also implements the new BIO read/write methods that take `size_t` lengths instead of `int`. This skips the compatibility methods provided by OpenSSL.

NOTE: the patch also starts to use uppercased external symbols, such as `BIO_set_init`, to avoid `bio_set_init = BIO_set_init` aliases. This is limited to the `BIO_` and `BIO_meth_*` used in this patch.

### CODE REVIEW

File: src/openssl/bio.cr [hunk 1]
```diff
@@ -34,31 +47,56 @@ struct OpenSSL::BIO
...
 
-    crystal_bio.create = LibCrypto::BioMethodCreate.new do |bio|
-      bio.value.shutdown = 1
-      bio.value.init = 1
-      bio.value.num = -1
+    create = LibCrypto::BioMethodCreate.new do |bio|
+      {% if LibCrypto::OPENSSL_110 %}
+        LibCrypto.BIO_set_shutdown(bio, 1)
+        LibCrypto.BIO_set_init(bio, 1)
+        # bio.value.num = -1
```
Comment (ysbaddaden) [inline src/openssl/bio.cr]: I have no idea what `BIO#num` means and what we should do here. Anybody knows?

Comment (bcardiff) [reply src/openssl/bio.cr]: It seems to come originally from https://github.com/crystal-lang/crystal/commit/4995fb010316ac0cb3085ab18f8738f7b7415140#diff-65c7a3575f24669104f83007f62e90c6R109 if the documentation of 1.1.x says nothing about it it's fine as it is I guess.

**Comment by bcardiff** [comment]
This kind of issues makes me want to enforce more infrastructure to test compatibility with different libraries. Otherwise we in lack of sight.

**Comment by RX14** [comment]
@bcardiff we'd have had to keep track of the openssl release and add it to the test infastructure. It's honestly better to be reactionary about these things to avoid wasting effort. Maybe once there's far more manpower in the project.

**Comment by RX14** [comment]
tested locally

**Comment by ysbaddaden** [comment]
What's really needed is a manual review of the OpenSSL integration which dates back to the old OpenSSL 0.9.8. See what public API they provide now, but was kinda internal/private.

For example the changes here were already present in OpenSSL 1.1.0, and would have avoided the reported segfault.

But that requires having an active maintainer...

**Pull request merged**